### PR TITLE
Support Android flavors in new architecture

### PR DESCRIPTION
**Pull request by oblador**
## Summary

When enabling the new architecture on Android apps using flavors the build fails with `Could not get unknown property 'preDebugBuild'` as reported in https://github.com/facebook/react-native/issues/33596. 

This PR fixes this by iterating over variants instead and dynamically applying the necessary dependencies. 

## Changelog

[Android] [Fixed] - Support Android flavors in new architecture

## Test Plan

Applied these changes to our Android project with multiple flavors and new architecture flag enabled.

### CODE REVIEW

File: template/android/app/build.gradle [hunk 1]
```diff
@@ -190,20 +190,25 @@ android {
             // If you wish to add a custom TurboModule or component locally,
             // you should uncomment this line.
             // preBuild.dependsOn("generateCodegenArtifactsFromSchema")
-            preDebugBuild.dependsOn(packageReactNdkDebugLibs)
-            preReleaseBuild.dependsOn(packageReactNdkReleaseLibs)
+            android.applicationVariants.all { def variant ->
```
Comment (cortinico) [inline template/android/app/build.gradle:193]: Why is this needed? The code was originally written under the assumption that you had both debug & release build type (which is safe to assume, as those are the default build types).

Are you removing one of the two?

**Review by cortinico — COMMENTED** [review]
Thanks for the PR @oblador 

I've left a question as I'd like to understand why this is needed in the first place 👍

File: template/android/app/build.gradle (hunk 1 above)
Comment (oblador) [reply template/android/app/build.gradle:193]: I'm not, but I've pushed a commit fixing a bug with the arch specific ones. Basic idea here is that a flavorless app typically has `preDebugBuild` & `preReleaseBuild` whereas an app with flavors has `pre$FlavorDebugBuild` & `pre$FlavorReleaseBuild`.

**Comment by SparshaSaha** [comment]
I can see that CMake installation fails with a 500 error code. We might consider re-running this step. I did not find any specific code changes in this PR which could fail that step.

File: template/android/app/build.gradle [hunk 2]
```diff
@@ -190,20 +190,30 @@ android {
...
-            configureNdkBuildDebug.dependsOn(preDebugBuild)
-            reactNativeArchitectures().each { architecture ->
-                tasks.findByName("configureNdkBuildDebug[${architecture}]")?.configure {
-                    dependsOn("preDebugBuild")
+                // Due to a bug inside AGP, we have to explicitly set a dependency
+                // between configureNdkBuild* tasks and the preBuild tasks.
+                // This can be removed once this is solved: https://issuetracker.google.com/issues/207403732
+                if (variantName.endsWith("Debug")) {
+                    preVariantBuildTask.dependsOn(packageReactNdkDebugLibs)
+                    configureNdkBuildDebug.dependsOn(preVariantBuildTask)
```
Comment (DomiR) [inline template/android/app/build.gradle:202]:
```suggestion
                    configureCMakeDebug.dependsOn(preVariantBuildTask)
```

File: template/android/app/build.gradle [hunk 3]
```diff
@@ -190,20 +190,30 @@ android {
...
+                    reactNativeArchitectures().each { architecture ->
+                        tasks.findByName("configureNdkBuildDebug[${architecture}]")?.configure {
```
Comment (DomiR) [inline template/android/app/build.gradle:204]:
```suggestion
                        tasks.findByName("configureCMakeDebug[${architecture}]")?.configure {
```

File: template/android/app/build.gradle [hunk 4]
```diff
@@ -190,20 +190,30 @@ android {
...
+                            dependsOn(preVariantBuildTask.name)
+                        }
+                    }
                 }
-                tasks.findByName("configureNdkBuildRelease[${architecture}]")?.configure {
-                    dependsOn("preReleaseBuild")
+                if (variantName.endsWith("Release")) {
+                    preVariantBuildTask.dependsOn(packageReactNdkReleaseLibs)
+                    configureNdkBuildRelease.dependsOn(preVariantBuildTask)
```
Comment (DomiR) [inline template/android/app/build.gradle:211]:
```suggestion
                    configureCMakeRelWithDebInfo.dependsOn(preVariantBuildTask)
```

File: template/android/app/build.gradle [hunk 5]
```diff
@@ -190,20 +190,30 @@ android {
...
+                    reactNativeArchitectures().each { architecture ->
+                        tasks.findByName("configureNdkBuildRelease[${architecture}]")?.configure {
```
Comment (DomiR) [inline template/android/app/build.gradle:213]:
```suggestion
                        tasks.findByName("configureCMakeRelWithDebInfo[${architecture}]")?.configure {
```

**Review by DomiR — COMMENTED** [review]
Now that 0.70.0 is out this can be updated to the new cmake build. Sucks, that this was not already merged 😢

**Comment by cortinico** [comment]
> Now that 0.70.0 is out this can be updated to the new cmake build. Sucks, that this was not already merged 😢 But thanks for this solution!
> 
> Edit: I guess both (`configureNdkBuildDebug` and `configureCMakeDebug`) are needed until libraries all migrate. So I'm deleting my change suggestions for now. I'm stuck building `vision-camera` and will try to report back as soon as I manage to build my project properly.
> Now that 0.70.0 is out this can be updated to the new cmake build. Sucks, that this was not already merged 😢 But thanks for this solution!
> 
> Edit: I guess both (`configureNdkBuildDebug` and `configureCMakeDebug`) are needed until libraries all migrate. So I'm deleting my change suggestions for now. I'm stuck building `vision-camera` and will try to report back as soon as I manage to build my project properly.

I'm going to close this PR as it's not needed anymore thanks to AGP 7.3:
https://github.com/facebook/react-native/pull/34707

We're going to get rid of all the `dependsOn` directive that users had to specify in their `build.gradle`.

We might consider backporting the AGP bump to 0.70.x if there is a high demand for it.

**Comment by ibnukipa** [comment]
@cortinico is there any plan to bump this one? or is there any plan to release 0.71.0?

**Comment by kelset** [comment]
@ibnukipa  AGP 7.3 will be released with 0.71

**Comment by ibnukipa** [comment]
@kelset great! thanks for confirming
